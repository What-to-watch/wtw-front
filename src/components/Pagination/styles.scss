.pagination {
    display: flex;
    justify-content: space-between;
    align-items: center;
    height: 68px;
    width: 290px;
    &__description {
        width: 150px;
        display: flex;
        justify-content: center;
        & span {
            color: white;
            font-weight: normal;
            font-size: 16px;
            &:last-child {
                margin-left: 5px;
            }
        }
    }
    & button {
        display: flex;
        justify-content: space-between;
        opacity: 1;
        & img:first-child {
            margin-right: 6px;
        }
    }

    & button[disabled], & button:disabled {
        opacity: 0.5;
        cursor: not-allowed;
    }
}