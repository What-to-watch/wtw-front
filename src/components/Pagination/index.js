import React from 'react';
import PropTypes from 'prop-types';

import './styles.scss';

const Pagination = (props) => {
    const {from, to, total, loading, onChangePage = x => x} = props;

    const handlePage = (type) => () => {
        onChangePage(type, props)
    }
    
    return (
        <div className="pagination">
            <button onClick={handlePage("first")} disabled={from === 1 || loading}>
                <img src="icons/pagination-line.svg" alt="|"/>
                <img src="icons/left-arrow.svg" alt="<"/>
            </button>
            <button onClick={handlePage("previous")} disabled={from === 1 || loading}>
                <img src="icons/left-arrow.svg" alt="<"/>
            </button>
            <div className="pagination__description">
                <span>{from}</span>
                <span>-</span>
                <span>{to}</span>
                <span> of {total} </span>
            </div>
            <button onClick={handlePage("next")} disabled={to === total || loading}>
                <img src="icons/right-arrow.svg" alt=">"/>
            </button>
            <button onClick={handlePage("last")} disabled={to === total || loading}>
                <img src="icons/right-arrow.svg" alt=">"/>
                <img src="icons/pagination-line.svg" alt="|"/>
            </button>
        </div>
    )
}

Pagination.propTypes = {
    from: PropTypes.number,
    to: PropTypes.number,
    total: PropTypes.number,
    loading: PropTypes.bool,
    onChangePage: PropTypes.func,
};

export default Pagination;