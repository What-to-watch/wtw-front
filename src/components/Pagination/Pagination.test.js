import React from 'react';
import Pagination from './index';
import { render, fireEvent, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

describe ('Pagination render', ()=>{
    let props = {
        page: 1,
        pages: 5,
        total: 450,
        perPage: 100,
        length: 100,
    }
    let component = renderer.create(
        (<Pagination {...props}/>),
    );

    let pagination = component.toJSON();

    it('Pagination render first page', ()=>{

        const first = ( props.perPage * props.page ) - props.perPage + 1;
        const last = first + props.length - 1;

        const disabledButtons = component.root.findAllByProps({disabled: true});
        const descriptionText = component.root.findAllByType('span');

        expect(disabledButtons.length).toBe(2);
        expect(descriptionText[0].children[0]).toBe(first.toString())
        expect(descriptionText[2].children[0]).toBe(last.toString())

        expect(pagination).toMatchSnapshot();

    });

    it('Pagination render last page', () => {
        props = { ...props, page: 5 } 
        component = renderer.create(
            (<Pagination {...props}/>),
        );

        pagination = component.toJSON();

        const first = ( props.perPage * props.page ) - props.perPage + 1;
        const last = first + props.length - 1;

        const disabledButtons = component.root.findAllByProps({disabled: true});
        const descriptionText = component.root.findAllByType('span');

        expect(disabledButtons.length).toBe(2);
        expect(descriptionText[0].children[0]).toBe(first.toString())
        expect(descriptionText[2].children[0]).toBe(last.toString())

        expect(pagination).toMatchSnapshot();
    })

    it('Pagination render in between page', () => {
        props = { ...props, page: 3 } 
        component = renderer.create(
            (<Pagination {...props}/>),
        );

        pagination = component.toJSON();

        const first = ( props.perPage * props.page ) - props.perPage + 1;
        const last = first + props.length - 1;

        const disabledButtons = component.root.findAllByProps({disabled: false});
        const descriptionText = component.root.findAllByType('span');

        expect(disabledButtons.length).toBe(4);
        expect(descriptionText[0].children[0]).toBe(first.toString())
        expect(descriptionText[2].children[0]).toBe(last.toString())

        expect(pagination).toMatchSnapshot();
    })
    it('Pagination ChangePage is executed with button', ()=>{
        const spy = (() => {
            let _spy = (...args) => {
                _spy.calls.push({ args })
                _spy.called = true;
                _spy.callCount = _spy.callCount+1;
            }
            _spy.calls = [];
            _spy.called = false;
            _spy.callCount = 0;
            Object.defineProperty(_spy,"firstCall",{
                get: () => _spy.calls[0]
            })
            return _spy
        })()

        props = { ...props, changePage: spy }

        render(<Pagination {...props} />);

        fireEvent.click(screen.getAllByRole('button')[0]);

        expect(spy.called).toBeTruthy();
        expect(spy.callCount).toBe(1);
        expect(spy.firstCall.args[0]).toBe(1);
    });

})