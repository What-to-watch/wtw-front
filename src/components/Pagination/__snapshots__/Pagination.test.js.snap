// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination render Pagination render first page 1`] = `
<div
  className="pagination"
>
  <button
    disabled={true}
    onClick={[Function]}
  >
    <img
      alt="|"
      src="icons/pagination-line.svg"
    />
    <img
      alt="<"
      src="icons/left-arrow.svg"
    />
  </button>
  <button
    disabled={true}
    onClick={[Function]}
  >
    <img
      alt="<"
      src="icons/left-arrow.svg"
    />
  </button>
  <div
    className="pagination__description"
  >
    <span>
      1
    </span>
    <span>
      -
    </span>
    <span>
      100
    </span>
    <span>
       of 
      450
       
    </span>
  </div>
  <button
    disabled={false}
    onClick={[Function]}
  >
    <img
      alt=">"
      src="icons/right-arrow.svg"
    />
  </button>
  <button
    disabled={false}
    onClick={[Function]}
  >
    <img
      alt=">"
      src="icons/right-arrow.svg"
    />
    <img
      alt="|"
      src="icons/pagination-line.svg"
    />
  </button>
</div>
`;

exports[`Pagination render Pagination render in between page 1`] = `
<div
  className="pagination"
>
  <button
    disabled={false}
    onClick={[Function]}
  >
    <img
      alt="|"
      src="icons/pagination-line.svg"
    />
    <img
      alt="<"
      src="icons/left-arrow.svg"
    />
  </button>
  <button
    disabled={false}
    onClick={[Function]}
  >
    <img
      alt="<"
      src="icons/left-arrow.svg"
    />
  </button>
  <div
    className="pagination__description"
  >
    <span>
      201
    </span>
    <span>
      -
    </span>
    <span>
      300
    </span>
    <span>
       of 
      450
       
    </span>
  </div>
  <button
    disabled={false}
    onClick={[Function]}
  >
    <img
      alt=">"
      src="icons/right-arrow.svg"
    />
  </button>
  <button
    disabled={false}
    onClick={[Function]}
  >
    <img
      alt=">"
      src="icons/right-arrow.svg"
    />
    <img
      alt="|"
      src="icons/pagination-line.svg"
    />
  </button>
</div>
`;

exports[`Pagination render Pagination render last page 1`] = `
<div
  className="pagination"
>
  <button
    disabled={false}
    onClick={[Function]}
  >
    <img
      alt="|"
      src="icons/pagination-line.svg"
    />
    <img
      alt="<"
      src="icons/left-arrow.svg"
    />
  </button>
  <button
    disabled={false}
    onClick={[Function]}
  >
    <img
      alt="<"
      src="icons/left-arrow.svg"
    />
  </button>
  <div
    className="pagination__description"
  >
    <span>
      401
    </span>
    <span>
      -
    </span>
    <span>
      500
    </span>
    <span>
       of 
      450
       
    </span>
  </div>
  <button
    disabled={true}
    onClick={[Function]}
  >
    <img
      alt=">"
      src="icons/right-arrow.svg"
    />
  </button>
  <button
    disabled={true}
    onClick={[Function]}
  >
    <img
      alt=">"
      src="icons/right-arrow.svg"
    />
    <img
      alt="|"
      src="icons/pagination-line.svg"
    />
  </button>
</div>
`;
